Lib/unittest/case.py

```diff
@@ -1216,20 +1216,38 @@ def assertCountEqual(self, first, second, msg=None):
             self.fail(msg)
 
     def assertMultiLineEqual(self, first, second, msg=None):
-        """Assert that two multi-line strings are equal."""
-        self.assertIsInstance(first, str, 'First argument is not a string')
-        self.assertIsInstance(second, str, 'Second argument is not a string')
+        """Assert that two multi-line strings are equal.
+         If the assertion fails, then provide an error message with a detailed
+         diff output
+        """
+        self.assertIsInstance(first, str, "First argument is not a string")
+        self.assertIsInstance(second, str, "Second argument is not a string")
 
         if first != second:
-            # don't use difflib if the strings are too long
+            # Don't use difflib if the strings are too long
             if (len(first) > self._diffThreshold or
                 len(second) > self._diffThreshold):
                 self._baseAssertEqual(first, second, msg)
-            firstlines = first.splitlines(keepends=True)
-            secondlines = second.splitlines(keepends=True)
-            if len(firstlines) == 1 and first.strip('\r\n') == first:
-                firstlines = [first + '\n']
-                secondlines = [second + '\n']
+
+            # Append \n to both strings if either is missing the \n.
+            # This allows the final ndiff to show the \n difference. The
+            # exception here is if the string is empty, in which case no
+            # \n should be added
+            first_presplit = first
+            second_presplit = second
+            if first != '' and second != '':
+                if first[-1] != '\n' or second[-1] != '\n':
+                    first_presplit += '\n'
+                    second_presplit += '\n'
+            elif first == '' and second != '' and second[-1] != '\n':
+                second_presplit += '\n'
+            elif second == '' and first != '' and first[-1] != '\n':
+                first_presplit += '\n'
```

Comment:
I think taking advantage of the fact that we know `first` and `second` are strings here, and empty strings are falsy, makes this a bit easier to read. Also once we've excluded the `first and second` case, `second and not first` can be simplified to just `second`:

```suggestion
            if first and second:
                if first[-1] != '\n' or second[-1] != '\n':
                    first_presplit += '\n'
                    second_presplit += '\n'
            elif second and second[-1] != '\n':
                second_presplit += '\n'
            elif first and first[-1] != '\n':
                first_presplit += '\n'
```

Reply:
I see what you mean about  _first != '' and second != ''_ ; simplifying to _first and second_ makes sense.  

On lines 1243 and 1245, the removal if _first == ''_ and then _second == ''_ changes the logic in a way that.  In those sections of code, the newline should only be added to the string if, and only if, the other is an empty string.

Reply:
We've already eliminated the case of `first and second` where both are non empty; that would hit the first `if` clause and can't reach either `elif`. That means that `elif first` must imply `not second` and vice versa.

If you think I'm wrong, try it out and see if you can get it to fail any tests!

Reply:
Don't confuse _first == '' and second == ''_ with _first == '' or second == ''_; the condition on 1239 is specifically looking for where BOTH are empty strings, where as the conditions below it are looking for one OR the other to be empty strings. You make a good point about the test cases. The test cases need to check for this situation. I will confirm and add to them if necessary. I'm also thinking I need to add more explanatory comments to the code, as even I admit, the logic confusing.

Reply:
Negative, with a binary state of each of two variables (empty-string or not-empty-string), there are four possible combinations [(not-empty, not-empty),(empty, empty),(not-empty, empty),(empty,not-empty)]. The if on 1239 catches (not-empty, not-empty), the elif on 1243 catches (empty,not-empty) while the one on 1245 catches (not-empty, empty).

Reply:
> with a binary state of each of two variables (empty-string or not-empty-string), there are four possible combinations [(not-empty, not-empty),(empty, empty),(not-empty, empty),(empty,not-empty)].

Right. The `if` on 1239 catches `(not-empty, not-empty)`. The remaining possible states are `(not-empty, empty)`, `(empty, not-empty)`, and `(empty, empty)`. So if line 1243 has `elif second ...` as I propose, the only possible remaining state that could pass that check is `(empty, not-empty)`. So it is redundant to also check `... and not first` in the elif: `if second` already guarantees that `first` is empty, since we know they can't both be non-empty.

Reply:
You are ... correct. Sorry it took so long for me to see the light. I had my wires crossed about the elif functionality.

Reply:
No worries! :)